prow/cmd/hmac/main.go

```diff
@@ -0,0 +1,389 @@
+		// Copy over all existing tokens for that repo.
+		updatedTokenList = append(updatedTokenList, val...)
+	} else {
+		// Current webhook is possibly using global token so we need to promote that token to repo level.
+		globalTokens := c.currentHMACMap["*"]
+		updatedTokenList = append(updatedTokenList, github.HmacSecret{
+			Value: globalTokens[0].Value,
+			// Set CreatedAt as a time slightly before now, so that the token can be properly pruned in the end.
+			CreatedAt: time.Now().Add(-time.Hour),
+		})
+	}
+
+	updatedTokenList = append(updatedTokenList, github.HmacSecret{
+		Value: generatedToken, CreatedAt: time.Now()})
+	c.currentHMACMap[repo] = updatedTokenList
+	c.hmacMapForBatchUpdate[repo] = generatedToken
+
+	return nil
+}
+
+func (c *client) batchOnboardNewTokenForRepos() error {
+	for repo, generatedToken := range c.hmacMapForBatchUpdate {
+		// Update the github webhook to use new token.
+		o := ghhook.Options{
+			GitHubOptions: c.options.github,
+			Repos:         prowflagutil.NewStrings(repo),
+			HookURL:       c.options.hookUrl,
+			HMACValue:     generatedToken,
+			// Receive hooks for all the events.
+			Events:  prowflagutil.NewStrings(github.AllHookEvents...),
+			Confirm: true,
+		}
+		if err := o.Initialize(); err != nil {
+			return err
+		}
+
+		if err := o.HandleWebhookConfigChange(); err != nil {
+			// Log and skip to next one.
+			return fmt.Errorf("error updating webhook for repo %q: %v", repo, err)
+		}
+	}
+
+	return nil
+}
+
+// cleanup will do necessary cleanups after the token and webhook updates are done.
+func (c *client) cleanup() error {
+	// Prune old tokens from current config.
+	for repoName := range c.currentHMACMap {
+		c.pruneOldTokens(repoName)
+	}
+	// Update the secret.
+	if err := c.updateHmacTokenSecret(); err != nil {
+		return fmt.Errorf("error updating hmac tokens: %v", err)
+	}
+	return nil
+}
+
+// updateHmacTokenSecret saves given in-memory config to secret file used by prow cluster.
+func (c *client) updateHmacTokenSecret() error {
+	if c.options.dryRun {
+		logrus.Debug("dryrun option is enabled, updateHmacTokenSecret won't actually update the secret.")
+		return nil
+	}
+
+	secretContent, err := yaml.Marshal(&c.currentHMACMap)
+	if err != nil {
+		return fmt.Errorf("error converting hmac map to yaml: %v", err)
+	}
+	secret := &corev1.Secret{}
+	secret.Name = c.options.hmacTokenSecretName
+	secret.Namespace = c.options.hmacTokenSecretNamespace
+	secret.StringData = map[string]string{c.options.hmacTokenKey: string(secretContent)}
+	if _, err = c.kubernetesClient.CoreV1().Secrets(c.options.hmacTokenSecretNamespace).Update(secret); err != nil {
+		return fmt.Errorf("error updating the secret: %v", err)
+	}
+	return nil
+}
+
+// pruneOldTokens removes all but most recent token from token config.
+func (c *client) pruneOldTokens(repo string) {
+	tokens := c.currentHMACMap[repo]
+	if len(tokens) <= 1 {
+		logrus.Debugf("Token size for repo %q is %d, no need to prune", repo, len(tokens))
```

Comment:
nit: If we care about logging this case then we should be logging the opposite case too. e.g. "Pruned X tokens for repo foo/bar."
Also using `logrus.WithField` or `logrus.WithFields` is a better way to include parameters like the repo since it enables us to filter the logs on that parameter.

Reply:
Updated the logging messages. Thanks!